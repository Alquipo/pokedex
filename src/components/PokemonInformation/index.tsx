import { PokemonTypesProps } from 'components/PokemonCard'
import PokemonTypeIcon from 'components/PokemonTypeIcon'
import {
  convertDecimetersToFeet,
  convertHectogramsToFeet
} from 'utils/convertInfPokemon'

import { formatNameID } from 'utils/formatName'

import * as S from './styles'

export type PokemonInformationProps = {
  id: number
  height: number
  wight: number
  abilities: string[]
  types: PokemonTypesProps[]
}

const PokemonInformation = ({
  id,
  height,
  wight,
  abilities,
  types
}: PokemonInformationProps) => (
  <S.Wrapper>
    <S.WrapperRotate>
      <S.Table>
        <tbody>
          <tr>
            <td>ID</td>
            <td>#{formatNameID(id)}</td>
          </tr>
          <tr>
            <td>Height</td>
            <td>
              {height / 10}m {`(${convertDecimetersToFeet(height)})`}
            </td>
          </tr>
          <tr>
            <td>Wight</td>
            <td>
              {wight / 10}Kg {`( ${convertHectogramsToFeet(wight)} )`}
            </td>
          </tr>
          <tr>
            <td>Abilities</td>
            <td>
              <S.WrapperAbilities>
                {abilities.map((ability, index) => (
                  <S.ButtonAbilities
                    key={index}
                    role="button"
                    typeColor={types[0]}
                  >
                    {ability}
                  </S.ButtonAbilities>
                ))}
              </S.WrapperAbilities>
            </td>
          </tr>

          <tr>
            <td>Type</td>
            <td>
              <S.WrapperTypes>
                {types.map((type, index) => (
                  <S.Type key={index} typeColor={type}>
                    <span>{type}</span>
                    <PokemonTypeIcon type={type} />
                  </S.Type>
                ))}
              </S.WrapperTypes>
            </td>
          </tr>
        </tbody>
      </S.Table>
    </S.WrapperRotate>
  </S.Wrapper>
)

export default PokemonInformation
