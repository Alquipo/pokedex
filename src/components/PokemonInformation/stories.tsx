import { Story, Meta } from '@storybook/react/types-6-0'
import PokemonInformation, { PokemonInformationProps } from '.'

export default {
  title: 'PokemonPage/PokemonInformation',
  component: PokemonInformation,
  args: {
    id: 6,
    height: 17,
    wight: 905,
    abilities: ['blaze', 'solar-power'],
    types: ['fire', 'flying']
  },
  argTypes: {
    types: {
      control: {
        type: 'inline-check',
        options: [
          'normal',
          'fire',
          'water',
          'electric',
          'grass',
          'ice',
          'ground',
          'flying',
          'ghost',
          'rock',
          'fighting',
          'poison',
          'psychic',
          'bug',
          'dark',
          'steel',
          'dragon',
          'fairy'
        ]
      }
    }
  }
} as Meta

export const Default: Story<PokemonInformationProps> = (args) => (
  <div
    style={{
      margin: '7rem auto 0 auto'
    }}
  >
    <PokemonInformation {...args} />
  </div>
)
