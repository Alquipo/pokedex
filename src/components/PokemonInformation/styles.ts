import { PokemonTypesProps } from 'components/PokemonCard'
import styled, { css } from 'styled-components'
import { generateMedia } from 'styled-media-query'

const customMedia = generateMedia({
  desktop: '980px'
})

type PokemonTypeColors = {
  typeColor: PokemonTypesProps
}

export const Wrapper = styled.section`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;
  padding-left: 2rem;

  ${customMedia.greaterThan('desktop')`
   perspective: 30rem;
  `}

  ${customMedia.lessThan('desktop')`
   margin-top: 2rem;
  `}

  cursor: default;
`

export const WrapperRotate = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;

  ${customMedia.greaterThan('desktop')`
      transform: rotateY(30deg);
      transition: 0.4s;
      &:hover {
        transform: rotateY(0);
      }
    `}
`

export const Table = styled.table`
  ${({ theme }) => css`
    line-height: 2;
    color: ${theme.colors.gray};

    margin-left: 4rem;
    border-collapse: collapse;

    font-size: ${theme.font.sizes.large};

    td {
      padding: 0.5rem;
    }

    td:first-child {
      font-weight: ${theme.font.medium};
      text-align: right;
      padding-right: 2rem;
    }

    tr:nth-child(4) {
      span:nth-child(2) {
        filter: brightness(110%) saturate(50%);
      }
    }
  `}
`

export const WrapperAbilities = styled.span`
  width: 27rem;
  display: flex;
  flex-wrap: wrap;
`

export const ButtonAbilities = styled.span<PokemonTypeColors>`
  ${({ theme, typeColor }) => css`
    color: ${theme.colors.white};
    font-weight: ${theme.font.medium};

    background-color: ${theme.colors.pokemonText[typeColor]};
    transition: ${theme.transition.backgroundColor};

    border-radius: ${theme.border.radius.xsmall};
    border: none;
    outline: none;

    padding: 0 0.3rem;
    margin-right: ${theme.spacings.xsmall};
    margin-bottom: ${theme.spacings.xsmall};

    text-transform: uppercase;

    cursor: pointer;
  `}
`

export const WrapperTypes = styled.div`
  display: flex;
  flex-direction: row;
`

export const Type = styled.div<PokemonTypeColors>`
  ${({ theme, typeColor }) => css`
    text-transform: capitalize;

    border-radius: ${theme.border.radius.small};
    height: 3rem;
    width: 9rem;

    margin-right: ${theme.spacings.small};

    display: flex;
    flex-direction: row;
    justify-content: space-between;

    background: ${theme.colors.pokemonType[typeColor]};
    box-shadow: 0 0 20px ${theme.colors.pokemonType[typeColor]};
    transition: all 0.2s;
    filter: saturate(100%) brightness(110%);

    cursor: default;

    span {
      line-height: 1;
      padding: 0.5rem 0 0 0.5rem;
      color: ${theme.colors.white};
      cursor: default;
    }

    div {
      margin: 0;
    }
  `}
`
