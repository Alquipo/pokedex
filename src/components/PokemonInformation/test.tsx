import { screen } from '@testing-library/react'

import PokemonInformation, { PokemonInformationProps } from '.'

import mockPokemon from 'templates/Pokemon/mock'
import { renderWithTheme } from 'utils/tests/helpers'

const props: PokemonInformationProps = {
  id: mockPokemon.id,
  height: mockPokemon.height,
  wight: mockPokemon.wight,
  abilities: mockPokemon.abilities,
  types: ['fire', 'flying']
}

describe('<PokemonInformation />', () => {
  it('should render the PokemonInformation correctly', () => {
    const { container } = renderWithTheme(<PokemonInformation {...props} />)

    expect(screen.getByRole('row', { name: /ID/i })).toBeInTheDocument()
    expect(screen.getByRole('row', { name: /Height/i })).toBeInTheDocument()
    expect(screen.getByRole('row', { name: /Wight/i })).toBeInTheDocument()
    expect(screen.getByRole('row', { name: /Abilities/i })).toBeInTheDocument()
    expect(screen.getByRole('row', { name: /Type/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
