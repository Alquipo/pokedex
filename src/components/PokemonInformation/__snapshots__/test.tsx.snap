// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonInformation /> should render the PokemonInformation correctly 1`] = `
.c7 {
  border-radius: 100%;
  height: 3rem;
  width: 3rem;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  margin: 2px 3px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FBA54C;
  box-shadow: 0 0 2rem #FBA54C;
}

.c9 {
  border-radius: 100%;
  height: 3rem;
  width: 3rem;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  margin: 2px 3px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #A1BBEC;
  box-shadow: 0 0 2rem #A1BBEC;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding-left: 2rem;
  cursor: default;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c2 {
  line-height: 2;
  color: #505050;
  margin-left: 4rem;
  border-collapse: collapse;
  font-size: 1.8rem;
}

.c2 td {
  padding: 0.5rem;
}

.c2 td:first-child {
  font-weight: 600;
  text-align: right;
  padding-right: 2rem;
}

.c2 tr:nth-child(4) span:nth-child(2) {
  -webkit-filter: brightness(110%) saturate(50%);
  filter: brightness(110%) saturate(50%);
}

.c3 {
  width: 27rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c4 {
  color: #fff;
  font-weight: 600;
  background-color: #ff8a80;
  -webkit-transition: background-color 1000ms ease-in,width 1500ms ease-in-out;
  transition: background-color 1000ms ease-in,width 1500ms ease-in-out;
  border-radius: 0.3rem;
  border: none;
  outline: none;
  padding: 0 0.3rem;
  margin-right: 0.8rem;
  margin-bottom: 0.8rem;
  text-transform: uppercase;
  cursor: pointer;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  text-transform: capitalize;
  border-radius: 0.5rem;
  height: 3rem;
  width: 9rem;
  margin-right: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #FBA54C;
  box-shadow: 0 0 20px #FBA54C;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  -webkit-filter: saturate(100%) brightness(110%);
  filter: saturate(100%) brightness(110%);
  cursor: default;
}

.c6 span {
  line-height: 1;
  padding: 0.5rem 0 0 0.5rem;
  color: #fff;
  cursor: default;
}

.c6 div {
  margin: 0;
}

.c8 {
  text-transform: capitalize;
  border-radius: 0.5rem;
  height: 3rem;
  width: 9rem;
  margin-right: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background: #A1BBEC;
  box-shadow: 0 0 20px #A1BBEC;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  -webkit-filter: saturate(100%) brightness(110%);
  filter: saturate(100%) brightness(110%);
  cursor: default;
}

.c8 span {
  line-height: 1;
  padding: 0.5rem 0 0 0.5rem;
  color: #fff;
  cursor: default;
}

.c8 div {
  margin: 0;
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:980px) {
  .c0 {
    -webkit-perspective: 30rem;
    -moz-perspective: 30rem;
    -ms-perspective: 30rem;
    perspective: 30rem;
  }
}

@media (max-width:980px) {
  .c0 {
    margin-top: 2rem;
  }
}

@media (min-width:980px) {
  .c1 {
    -webkit-transform: rotateY(30deg);
    -ms-transform: rotateY(30deg);
    transform: rotateY(30deg);
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }

  .c1:hover {
    -webkit-transform: rotateY(0);
    -ms-transform: rotateY(0);
    transform: rotateY(0);
  }
}

<section
  class="c0"
>
  <div
    class="c1"
  >
    <table
      class="c2"
    >
      <tbody>
        <tr>
          <td>
            ID
          </td>
          <td>
            #
            006
          </td>
        </tr>
        <tr>
          <td>
            Height
          </td>
          <td>
            0.16999999999999998
            m 
            (0.6ft)
          </td>
        </tr>
        <tr>
          <td>
            Wight
          </td>
          <td>
            9
            Kg 
            ( 19.8lbs )
          </td>
        </tr>
        <tr>
          <td>
            Abilities
          </td>
          <td>
            <span
              class="c3"
            >
              <span
                class="c4"
                role="button"
              >
                blaze
              </span>
              <span
                class="c4"
                role="button"
              >
                solar-power
              </span>
            </span>
          </td>
        </tr>
        <tr>
          <td>
            Type
          </td>
          <td>
            <div
              class="c5"
            >
              <div
                class="c6"
              >
                <span>
                  fire
                </span>
                <div
                  class="c7"
                >
                  <div
                    style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                  >
                    <div
                      style="box-sizing: border-box; display: block; max-width: 100%;"
                    >
                      <img
                        alt=""
                        aria-hidden="true"
                        role="presentation"
                        src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjIiIGhlaWdodD0iMjIiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                        style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                      />
                    </div>
                    <img
                      alt="fire"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="c8"
              >
                <span>
                  flying
                </span>
                <div
                  class="c9"
                >
                  <div
                    style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                  >
                    <div
                      style="box-sizing: border-box; display: block; max-width: 100%;"
                    >
                      <img
                        alt=""
                        aria-hidden="true"
                        role="presentation"
                        src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjIiIGhlaWdodD0iMjIiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                        style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                      />
                    </div>
                    <img
                      alt="flying"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                  </div>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
